articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,188 @@
+author: '[Nick Burek](https://github.com/nburek)'
+published: true
+categories: Middleware
+date: February 13th 2019
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+Glossary:
+
+- DDS - Data Distribution Service
+- RTPS - Real-Time Publish Subscribe
+- QoS - Quality of Service
+
+## Existing ROS QoS Settings
+
+While DDS provides many settings to enable fine-grained control over the Quality of Service (QOS) for entities, ROS only provides native support for a handful of them. ROS users are able to specify the History, Depth, Reliability, and Durability via a QoS configuration struct when they create Publishers, Subscribers, etc.
+
+This leaves a lot of QoS settings that can only be set if DDS vendor can load additional default settings via a configuration file.
+If a user wants to hook their code into these additional QoS settings then they would need to get a reference to the rmw implementation and program against the vendor specific API.
+Without the abstraction layer provided by ROS their code becomes less portable.
+
+## Proposed Changes
+
+As users start to build more robust applications, they are going to need more control over when data is being delivered and information about when the system is failing.
+To address these needs it was proposed that we start by adding support for the following new DDS QoS settings.
+
+### Proposed New QoS Settings
+
+- Deadline - The deadline policy establishes a contract for how often publishers will receive data and how often subscribers expect to receive data.
+- Liveliness - The liveliness policy is useful for determining if entities on the network are "alive" and notifying interested parties when they are not.
+- Lifespan - The lifespan policy is useful for preventing the delivery of stale messages by specifying a time at which they should no longer be delivered.
+
+More detailed information on these policies can be found below in Appendix A.
+
+### DDS Vendor Support
+
+Some of the primary DDS vendors that ROS supports already provide support for these QoS settings.
+However, the default vendor, Fast-RTPS, does not yet provide full support for these new QoS settings in their library.
+This does not have to be a blocker for adding support for these policy settings to ROS.
+We can start adding support for them and just log warning messages when a user attempts to specify an unsupported QoS setting.
+When support is eventually added for the policies in the vendor library then we can enable it in the rmw implementation.
+
+### Native ROS Support vs Vendor Configuration
+
+It may be argued that many DDS QoS settings could be used with ROS by simply specifying them using the vendor specific configuration mechanism, such as the xml configuration files used by RTI Connext.
```

Comment:
The XML specification of QoS settings is covered by the DDS4CCM specification, and more recently has been unified with all the other DDS-related XML stuff from other specifications into the single [DDS-XML](https://www.omg.org/spec/DDS-XML/About-DDS-XML/) specification. The XML approach being a standard means that it is a valid approach for ROS2 to use for the *specification* of these settings. It won't cover application response, though.

Reply:
Interesting. I may add some of that additional information into my updates. There were other concerns brought up about how non-DDS based rmw layers would work with these new settings. With that in mind, I'd want to stick close to how existing QoS settings are being provided for now. 

A discussion on if ROS2 should adopt the DDS-XML spec for QoS would be interesting, but I think it's out of scope of what I'm trying to achieve here with just getting support for these three new policies.

Reply:
I don't necessarily think it should, because then non-DDS RMW developers would have to comply with a big non-ROS specification just to be usable with ROS. I do think that use of it should be an option when using a DDS RMW that supports that specification.

DDS vendors originally didn't have much support for setting things via XML and the API was the way things go. The feature was added because it's needed to tune systems without constantly recompiling stuff.